File: dynamic-programming/fibonacci/fibonacci.cpp
// Dynamic Programming: Fibonacci with memoization
#include <iostream>
#include <vector>
using namespace std;

long long fib(int n, vector<long long>& memo) {
    if (n <= 1) return n;
    if (memo[n] != -1) return memo[n];
    return memo[n] = fib(n - 1, memo) + fib(n - 2, memo);
}

int main() {
    int n = 20;
    vector<long long> memo(n + 1, -1);

    cout << "fib(" << n << ") = " << fib(n, memo) << endl;

    return 0;
}

File: dynamic-programming/fibonacci/fibonacci.ts
// Dynamic Programming: Fibonacci with memoization

const memo = new Map<number, number>();

function fib(n: number): number {
  if (n <= 1) return n;
  if (memo.has(n)) return memo.get(n)!;

  const result = fib(n - 1) + fib(n - 2);
  memo.set(n, result);
  return result;
}

function main(): void {
  const n = 20;
  console.log(`fib(${n}) = ${fib(n)}`);
}

main();
